### PR TITLE
Clean-up `System.Drawing.Common` and remove the Unix code.

### PR DESCRIPTION
**Pull request by teo-tsirpanis**
Since #64084 got merged, `System.Drawing.Common` is supported only on Windows. This PR actually removes all Unix-specific code and tests, and merges the `.Windows.cs` files into their formerly cross-platform counterparts, drastically simplifying the codebase.

I also cleaned-up the code a little bit (used a `[ThreadStatic]` instead of a named data slot, and removed unused files).

### CODE REVIEW

**Comment by safern** [comment]
Thanks, @teo-tsirpanis -- please let me know when this is ready for review (CI is Red). I'd like to ask if you need any help to get CI green or a review, or just wait?

File: src/libraries/System.Drawing.Common/src/System.Drawing.Common.csproj [hunk 1]
```diff
@@ -201,49 +201,30 @@ Unix support is disabled by default. See https://aka.ms/systemdrawingnonwindows
   <!-- Windows-specific -->
   <ItemGroup Condition="'$(TargetsWindows)' == 'true'">
```
Comment (teo-tsirpanis) [inline src/libraries/System.Drawing.Common/src/System.Drawing.Common.csproj:202]: I got confused over what to do with these items' conditions. This is always true and the same applies with the one on line 33? But we are also building editions that don't contain any code.

File: src/libraries/System.Drawing.Common/src/System.Drawing.Common.csproj [hunk 2]
```diff
@@ -334,63 +315,16 @@ Unix support is disabled by default. See https://aka.ms/systemdrawingnonwindows
...
-    <Compile Include="System\Drawing\Icon.Windows.COMWrappers.cs" />
+    <Compile Include="System\Drawing\Icon.COMWrappers.cs" />
     <Compile Include="System\Drawing\Internal\GPStream.COMWrappers.cs" />
   </ItemGroup>
   <ItemGroup Condition="'$(TargetsWindows)' == 'true' and '$(TargetFrameworkIdentifier)' != '.NETCoreApp'">
     <Compile Include="$(CommonPath)Interop\Windows\Ole32\Interop.IStream.NoCOMWrappers.cs"
              Link="Common\Interop\Windows\Ole32\Interop.IStream.NoCOMWrappers.cs" />
     <Compile Include="System\Drawing\DrawingCom.NoCOMWrappers.cs" />
-    <Compile Include="System\Drawing\Icon.Windows.NoCOMWrappers.cs" />
+    <Compile Include="System\Drawing\Icon.NoCOMWrappers.cs" />
```
Comment (teo-tsirpanis) [inline src/libraries/System.Drawing.Common/src/System.Drawing.Common.csproj:325]: Should I keep the `NoCOMWrappers` files? My understanding is that we ship real code only on .NET 6 and 7, both of which have `ComWrappers`.

**Review by teo-tsirpanis — COMMENTED** [review]
Thanks for offering to help @safern. CI failures are so far unrelated (they occur on non-Windows platforms). I have two questions.

File: src/libraries/System.Drawing.Common/src/System.Drawing.Common.csproj (hunk 1 above)
Comment (danmoseley) [reply src/libraries/System.Drawing.Common/src/System.Drawing.Common.csproj:202]: When building for `netstandard2.0`, `TargetsWindows` would be false. The assembly in this case will just every public member, each throwing PlatformNotSupportedException.

Comment (teo-tsirpanis) [reply src/libraries/System.Drawing.Common/src/System.Drawing.Common.csproj]: I see, and the `$(TargetsAnyOS) != true` above used to be true only on `net(6|7).0-(windows|unix)`, and now that Unix is gone, only Windows remains and the two groups can be merged. Done.

Comment (safern) [reply src/libraries/System.Drawing.Common/src/System.Drawing.Common.csproj]: Exactly.

File: src/libraries/System.Drawing.Common/src/System.Drawing.Common.csproj (hunk 2 above)
Comment (safern) [reply src/libraries/System.Drawing.Common/src/System.Drawing.Common.csproj:321]: I think we can remove NoCOMWrappers files.

File: src/libraries/System.Drawing.Common/src/System.Drawing.Common.csproj [hunk 3]
```diff
@@ -1,11 +1,10 @@
-﻿<Project Sdk="Microsoft.NET.Sdk">
+<Project Sdk="Microsoft.NET.Sdk">
   <PropertyGroup>
     <DefineConstants>$(DefineConstants);DRAWING_NAMESPACE</DefineConstants>
     <AllowUnsafeBlocks>true</AllowUnsafeBlocks>
     <WarningsNotAsErrors>CS0618</WarningsNotAsErrors>
-    <DefineConstants Condition="'$(TargetsWindows)' == 'true'">$(DefineConstants);FEATURE_WINDOWS_SYSTEM_COLORS;FEATURE_SYSTEM_EVENTS</DefineConstants>
     <IsPartialFacadeAssembly>true</IsPartialFacadeAssembly>
-    <TargetFrameworks>$(NetCoreAppCurrent)-windows;$(NetCoreAppCurrent)-Unix;$(NetCoreAppCurrent);$(NetCoreAppMinimum)-windows;$(NetCoreAppMinimum)-Unix;$(NetCoreAppMinimum);netstandard2.0;$(NetFrameworkMinimum)</TargetFrameworks>
+    <TargetFrameworks>$(NetCoreAppCurrent)-windows;$(NetCoreAppCurrent);$(NetCoreAppMinimum)-windows;$(NetCoreAppMinimum);netstandard2.0;$(NetFrameworkMinimum)</TargetFrameworks>
```
Comment (safern) [inline src/libraries/System.Drawing.Common/src/System.Drawing.Common.csproj:7]: This will cause System.Drawing.Common to throw PlatformNotSupportedException on non Windows on any code path. Whereas now, it is throwing only if libgdiplus is tried to load. @ericstj and I chatted about this and thought this might break a very edge scenario where something can be used without loading libgdiplus at all and some library using that scenario will now be broken. I think this would be very rare but something worth discussion; I think I'm supporting of throwing PNSE everywhere, but people might differ on going the aggressive path:  @danmoseley @ericstj @terrajobst thoughts?

File: src/libraries/System.Drawing.Common/src/System.Drawing.Common.csproj [hunk 4]
```diff
@@ -329,68 +299,28 @@ Unix support is disabled by default. See https://aka.ms/systemdrawingnonwindows
...
+    <EmbeddedResource Include="Resources\System\Drawing\DefaultComponent.bmp">
+      <LogicalName>System.Drawing.DefaultComponent.bmp</LogicalName>
+    </EmbeddedResource>
+    <EmbeddedResource Include="Resources\System\Drawing\Printing\PrintDocument.bmp" />
+    <EmbeddedResource Include="Resources\System\Drawing\ShieldIcon.ico">
+      <LogicalName>System.Drawing.ShieldIcon.ico</LogicalName>
+    </EmbeddedResource>
   </ItemGroup>
-  <ItemGroup Condition="'$(TargetsWindows)' == 'true' and '$(TargetFrameworkIdentifier)' == '.NETCoreApp'">
+  <ItemGroup Condition="'$(TargetFrameworkIdentifier)' == '.NETCoreApp'">
```
Comment (safern) [inline src/libraries/System.Drawing.Common/src/System.Drawing.Common.csproj:310]: This move it to the `targetswindows` block as well, .NET Framework build will never set that to true as it doesn't have a windows suffix on the tfm.

File: src/libraries/System.Drawing.Common/src/System.Drawing.Common.csproj (hunk 3 above)
Comment (danmoseley) [reply src/libraries/System.Drawing.Common/src/System.Drawing.Common.csproj:7]: What is useful in S.D.C on Unix that wouldn't require libgdiplus?

Comment (safern) [reply src/libraries/System.Drawing.Common/src/System.Drawing.Common.csproj:7]: I'd have to look at that. I can't think from top of my mind, but there might be some to just carry data (as data holders) without any data manipulation.

File: src/libraries/System.Drawing.Common/src/System.Drawing.Common.csproj (hunk 2 above)
Comment (teo-tsirpanis) [reply src/libraries/System.Drawing.Common/src/System.Drawing.Common.csproj]: Done.

File: src/libraries/System.Drawing.Common/src/System.Drawing.Common.csproj (hunk 4 above)
Comment (teo-tsirpanis) [reply src/libraries/System.Drawing.Common/src/System.Drawing.Common.csproj:315]: Done.

File: src/libraries/System.Drawing.Common/src/System.Drawing.Common.csproj (hunk 3 above)
Comment (teo-tsirpanis) [reply src/libraries/System.Drawing.Common/src/System.Drawing.Common.csproj:7]: On non-Windows platforms I set a dummy `DllImportResolver` that fails with that PNSE when a library is trying to be loaded from SDC.

**Comment by teo-tsirpanis** [comment]
Looks like #64279 and #64720 have brought quite the merge conflicts on this PR. It will take me some time until I can fix them, no reason to close it.

**Comment by ViktorHofer** [comment]
@teo-tsirpanis https://github.com/dotnet/runtime/pull/64500 which got just merged caused a conflict which I just resolved. Hope that's ok.

**Comment by teo-tsirpanis** [comment]
Thanks @ViktorHofer.

File: src/libraries/System.Drawing.Common/src/System.Drawing.Common.csproj (hunk 4 above)
Comment (teo-tsirpanis) [reply src/libraries/System.Drawing.Common/src/System.Drawing.Common.csproj]: I changed the condition to `'$(TargetPlatformIdentifier)' != ''`, the same with the source files. I prefer to keep the assembly and project references in a separate group for clarity and to minimize layout changes.

Comment (teo-tsirpanis) [reply src/libraries/System.Drawing.Common/src/System.Drawing.Common.csproj]: It doesn't work, CI failed. Haven't actually understood what you asked me to do. 😵‍💫

**Comment by teo-tsirpanis** [comment]
Conflicts are resolved.

**Comment by danmoseley** [comment]
@dotnet/area-system-drawing as @safern changed teams, this will need a different one of you to sign off.

**Comment by teo-tsirpanis** [comment]
Great, CI now passes. Can someone review?

`BinaryFormatter` on some enum types breaks on non-Windows because the `TypeForwardedFromAttribute` is missing, but I made these test cases run only on Windows. I hope there's no problem.

**Comment by danmoseley** [comment]
we will need a new reviewer with Santi gone.

File: src/libraries/System.Runtime.Serialization.Formatters/tests/BinaryFormatterTestData.cs [hunk 5]
```diff
@@ -1185,13 +1185,13 @@ public static IEnumerable<object[]> SerializableObjects()
...
+				yield return new object[] { ContentAlignment.BottomCenter, new TypeSerializableValue[] { new TypeSerializableValue("AAEAAAD/////AQAAAAAAAAAMAgAAAFFTeXN0ZW0uRHJhd2luZywgVmVyc2lvbj00LjAuMC4wLCBDdWx0dXJlPW5ldXRyYWwsIFB1YmxpY0tleVRva2VuPWIwM2Y1ZjdmMTFkNTBhM2EFAQAAAB9TeXN0ZW0uRHJhd2luZy5Db250ZW50QWxpZ25tZW50AQAAAAd2YWx1ZV9fAAgCAAAAAAIAAAs=", TargetFrameworkMoniker.netcoreapp21), new TypeSerializableValue("AAEAAAD/////AQAAAAAAAAAMAgAAAFFTeXN0ZW0uRHJhd2luZywgVmVyc2lvbj00LjAuMC4wLCBDdWx0dXJlPW5ldXRyYWwsIFB1YmxpY0tleVRva2VuPWIwM2Y1ZjdmMTFkNTBhM2EFAQAAAB9TeXN0ZW0uRHJhd2luZy5Db250ZW50QWxpZ25tZW50AQAAAAd2YWx1ZV9fAAgCAAAAAAIAAAs=", TargetFrameworkMoniker.netfx461) } };
```
Comment (danmoseley) [inline src/libraries/System.Runtime.Serialization.Formatters/tests/BinaryFormatterTestData.cs:1192]: nit, did tabs creep in? we use 4 space indents.

Comment (teo-tsirpanis) [reply src/libraries/System.Runtime.Serialization.Formatters/tests/BinaryFormatterTestData.cs:1192]: Fixed. Guess it was Notepad++ because VSCode could not handle that big of a file.

**Comment by carlossanlop** [comment]
Hey @danmoseley I'll help review this soon. Santi left some instructions, need to read them first.

**Comment by teo-tsirpanis** [comment]
@carlossanlop any news? I'm gonna need two hands to count how many times I have resolved merge conflicts on this one. 😝

**Comment by ViktorHofer** [comment]
@teo-tsirpanis sorry for the long baking time on our side. I will be the champion for this PR and will make sure that it gets forward asap.

File: src/libraries/System.Drawing.Common/src/System.Drawing.Common.csproj [hunk 6]
```diff
@@ -332,60 +308,15 @@ Unix support is disabled by default. See https://aka.ms/systemdrawingnonwindows
...
+    <EmbeddedResource Include="Resources\System\Drawing\DefaultComponent.bmp">
+      <LogicalName>System.Drawing.DefaultComponent.bmp</LogicalName>
+    </EmbeddedResource>
+    <EmbeddedResource Include="Resources\System\Drawing\Printing\PrintDocument.bmp" />
+    <EmbeddedResource Include="Resources\System\Drawing\ShieldIcon.ico">
+      <LogicalName>System.Drawing.ShieldIcon.ico</LogicalName>
     </EmbeddedResource>
   </ItemGroup>
-  <ItemGroup Condition="$([MSBuild]::IsTargetFrameworkCompatible('$(TargetFramework)', 'net7.0')) and '$(TargetPlatformIdentifier)' != ''">
+  <ItemGroup Condition="$([MSBuild]::IsTargetFrameworkCompatible('$(TargetFramework)', 'net7.0')) and '$(TargetsAnyOS)' != 'true'">
```
Comment (ViktorHofer) [inline src/libraries/System.Drawing.Common/src/System.Drawing.Common.csproj:319]: The `TargetsAnyOS` property doesn't exist anymore and the above version is less verbose and does the same job.

```suggestion
  <ItemGroup Condition="$([MSBuild]::IsTargetFrameworkCompatible('$(TargetFramework)', 'net7.0-windows'))">
```

File: src/libraries/System.Drawing.Common/src/System/Drawing/Bitmap.cs [hunk 7]
```diff
@@ -53,6 +54,26 @@ public Bitmap(Stream stream) : this(stream, false)
         {
         }
 
+        public unsafe Bitmap(Stream stream!!, bool useIcm)
```
Comment (ViktorHofer) [inline src/libraries/System.Drawing.Common/src/System/Drawing/Bitmap.cs:57]: Note that the `!!` is being removed with https://github.com/dotnet/runtime/pull/68178.

Please use `ArgumentNullException.ThrowIfNull(stream);` instead.

File: src/libraries/System.Drawing.Common/src/System/Drawing/Gdiplus.cs [hunk 8]
```diff
@@ -42,21 +45,7 @@ static Gdip()
...
-                    if (threadData == null)
-                    {
-                        threadData = new Hashtable();
-                        Thread.SetData(slot, threadData);
-                    }
-
-                    return threadData;
-                }
-            }
+            internal static IDictionary ThreadData => t_threadData ??= new Hashtable();
```
Comment (ViktorHofer) [inline src/libraries/System.Drawing.Common/src/System/Drawing/Gdiplus.cs:48]: Curious, why did you change this?

File: src/libraries/System.Drawing.Common/src/System/Drawing/Gdiplus.cs [hunk 9]
```diff
@@ -18,13 +15,19 @@ internal static partial class SafeNativeMethods
...
+            private static IDictionary? t_threadData;
 
             static Gdip()
             {
-                Debug.Assert(s_initToken == IntPtr.Zero, "GdiplusInitialization: Initialize should not be called more than once in the same domain!");
+                if (!OperatingSystem.IsWindows())
+                {
+                    NativeLibrary.SetDllImportResolver(Assembly.GetExecutingAssembly(), static (_, _, _) =>
+                        throw new PlatformNotSupportedException(SR.PlatformNotSupported_Unix));
+                }
```
Comment (ViktorHofer) [inline src/libraries/System.Drawing.Common/src/System/Drawing/Gdiplus.cs:28]: I didn't follow on the official announcement but I assume it's intentionally that we explicitly block customers from loading this assembly on Unix?

File: src/libraries/System.Drawing.Common/src/System/Drawing/Gdiplus.cs (hunk 8 above)
Comment (teo-tsirpanis) [reply src/libraries/System.Drawing.Common/src/System/Drawing/Gdiplus.cs:48]: `[ThreadStatic]` is faster and simpler, and named data slots are nowhere used in the BCL (the shared framework at least). I was originally going to add this commit to another PR that will come after this one, but moved it here.

**Review by ViktorHofer — COMMENTED** [review]
Went over all the changes. This looks REALLY good. Thanks for the great work. The remaining work now is to:

1. Switch all the `!!` null checks to their previous state. For most, you could probably use `ArgumentNullException.ThrowIfNull(myString);`.
2. Make sure that the remaining questions are answered.

After that we can get the PR merged :)

PS: I hope you don't mind that I pushed one commit to your branch to clean-up the SDC project file further.

File: src/libraries/System.Drawing.Common/src/System/Drawing/Gdiplus.cs (hunk 8 above)
Comment (ViktorHofer) [reply src/libraries/System.Drawing.Common/src/System/Drawing/Gdiplus.cs:48]: @stephentoub would you mind lending me your pair of eyes to make sure this change is fine?

File: src/libraries/System.Drawing.Common/src/System/Drawing/Gdiplus.cs (hunk 9 above)
Comment (teo-tsirpanis) [reply src/libraries/System.Drawing.Common/src/System/Drawing/Gdiplus.cs:28]: That was the original plan. I first wrote

``` csharp
if (!OperatingSystem.IsWindows())
{
    throw new PlatformNotSupportedException(SR.PlatformNotSupported_Unix);
}
```

on the static constructor, which would throw more aggressively on non-Windows, but [after feedback](https://github.com/dotnet/runtime/pull/64623#pullrequestreview-869971017) I made it throw only if a P/Invoke from the SDC assembly was going to be called, which is the actual thing that is not supported on Unix, in the strictest sense. I don't think that just loading the assembly on Unix will cause any failures.

File: src/libraries/System.Drawing.Common/src/System/Drawing/Bitmap.cs (hunk 7 above)
Comment (teo-tsirpanis) [reply src/libraries/System.Drawing.Common/src/System/Drawing/Bitmap.cs:57]: I am working on this.

File: src/libraries/System.Drawing.Common/src/System/Drawing/Gdiplus.cs (hunk 8 above)
Comment (stephentoub) [reply src/libraries/System.Drawing.Common/src/System/Drawing/Gdiplus.cs:48]: Looks fine, thanks.

File: src/libraries/System.Drawing.Common/src/System/Drawing/Bitmap.cs (hunk 7 above)
Comment (teo-tsirpanis) [reply src/libraries/System.Drawing.Common/src/System/Drawing/Bitmap.cs]: Done.

**Review by ViktorHofer — APPROVED** [review]
Congratulations and thank you for your contribution 🎉🎉🎉 We very much appreciate the time and effort that you put into this change.

**Comment by ViktorHofer** [comment]
If you want to pick up another task, there are plenty ones in the backlog that are marked as up-for-grabs. And I will promise that future changes won't take us that long to react and will hopefully cause less merge conflicts (😥) on your side.

**Comment by danmoseley** [comment]
I also wanted to add my belated thanks here @teo-tsirpanis. This was a lot of work.
